### PR TITLE
[SPARK-41775][PYTHON][ML] Adding support for PyTorch functions

### PR DESCRIPTION
**Pull request by rithwik-db**
NOTE: If you want to only view the diff from the other WIP PR regarding the baseline changes, look at the LAST COMMIT in this PR's commit history (titled "WIP adding notebook functionality"). Since I am sending out parallel PRs that are related, you should view this commit to see the diff pertaining to this ticket.

### What changes were proposed in this pull request?

This is an addition to https://github.com/apache/spark/pull/39299 to add support for using functions as the input for distributed training. The users would follow the first workflow in the [design document](https://docs.google.com/document/d/1QPO1Ly8WteL6aIPvVcR7Xne9qVtJiB3fdrRn7NwBcpA/edit#heading=h.8yvw9xq428fh) to run training.

### Why are the changes needed?

We want to make it easier for users to run distributed training from a notebook setting.

### Does this PR introduce _any_ user-facing change?

Users can now input a training function into the `PyTorchDistributor().run(...)` api. This will require a lot of additional documentation though since since we are internally using cloudpickle to run training so we need to be able to have the user's train() function be picklable.  

### How was this patch tested?

This is a WIP PR so it hasn't been tested yet.

### CODE REVIEW

File: python/pyspark/ml/torch/distributor.py [hunk 1]
```diff
@@ -574,6 +582,68 @@ def _run_training_on_pytorch_file(
...
+        TorchDistributor._run_training_on_pytorch_file(input_params, train_file_path, *args)
+
+        output = TorchDistributor._get_pickled_output(output_file_path)
+        TorchDistributor._cleanup_files(save_dir)
+        return output
+
+    @staticmethod
+    def _create_save_dir() -> str:
+        # TODO: need to do this in a safe way to avoid issues during concurrent runs
+        return tempfile.mkdtemp()
```
Comment (WeichenXu123) [inline python/pyspark/ml/torch/distributor.py:612]: pls delete it when training completes or raises error.
you can convert it to a python context manager

File: python/pyspark/ml/torch/distributor.py [hunk 2]
```diff
@@ -574,6 +582,68 @@ def _run_training_on_pytorch_file(
...
+
+    @staticmethod
+    def _cleanup_files(save_dir: str) -> None:
```
Comment (WeichenXu123) [inline python/pyspark/ml/torch/distributor.py:615]: The method is not called when training routine raises error.

**Comment by WeichenXu123** [comment]
overall good.

File: python/pyspark/ml/torch/distributor.py [hunk 3]
```diff
@@ -495,32 +546,119 @@ def set_gpus(context: "BarrierTaskContext") -> None:
...
+        shutil.rmtree(save_dir)
```
Comment (WeichenXu123) [inline python/pyspark/ml/torch/distributor.py:626]: pls add argument `ignore_errors=True`

**Review by WeichenXu123 — APPROVED** [review]

File: python/pyspark/ml/torch/tests/test_distributor.py [hunk 4]
```diff
@@ -224,8 +293,10 @@ def setUp(self) -> None:
 
         self.sc = SparkContext("local-cluster[2,2,1024]", class_name, conf=conf)
         self.spark = SparkSession(self.sc)
+        self.mnist_dir_path = tempfile.mkdtemp()
 
     def tearDown(self) -> None:
+        shutil.rmtree(self.mnist_dir_path)
         os.unlink(self.tempFile.name)
```
Comment (WeichenXu123) [inline python/pyspark/ml/torch/tests/test_distributor.py:300]: Nit: refine the variable name: `tempFile` --> `gpu_discovery_script_file`

File: python/pyspark/ml/torch/tests/test_distributor.py [hunk 5]
```diff
@@ -349,6 +434,13 @@ def test_get_num_tasks_distributed(self) -> None:
 
         self.spark.sparkContext._conf.set("spark.task.resource.gpu.amount", "1")
 
+    # def test_end_to_end_run_distributedly(self) -> None:
```
Comment (lu-wang-dl) [inline python/pyspark/ml/torch/tests/test_distributor.py:437]: Clean up this?

Comment (rithwik-db) [reply python/pyspark/ml/torch/tests/test_distributor.py:437]: Sorry, that shouldn't have been commented out. Will quickly update adding that test back in

**Review by lu-wang-dl — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

File: python/pyspark/ml/torch/distributor.py [hunk 6]
```diff
@@ -15,16 +15,21 @@
 # limitations under the License.
 #
 
+import cloudpickle  # type: ignore
```
Comment (dongjoon-hyun) [inline python/pyspark/ml/torch/distributor.py:18]: This should be our `cloudpickle`.

**Review by dongjoon-hyun — COMMENTED** [review]
Hi, @rithwik-db , @WeichenXu123 , @HyukjinKwon .
Apache Spark should not require two versions of `cloudpickle` at the same time.
I made a PR, #39715.

File: python/pyspark/ml/torch/distributor.py (hunk 6 above)
Comment (HyukjinKwon) [reply python/pyspark/ml/torch/distributor.py:18]: +1 thanks!

**Pull request closed without merging**